packages/core/src/components/axes/toolbar.ts

```diff
@@ -0,0 +1,495 @@
+
+// D3 Imports
+import { event, select } from "d3-selection";
+
+// import the settings for the css prefix
+import settings from "carbon-components/es/globals/js/settings";
+
+export class Toolbar extends Component {
+	type = "toolbar";
+
+	// overflow menu button to control background color
+	overflowButton: any;
+
+	// ul options list element
+	overflowMenu: any;
+
+	// x, y coordinate of overflow menu
+	overflowMenuX = 0;
+	overflowMenuY = 0;
+
+	// Use a random number to create overflow menu item unique ID
+	// so they don't interfere the other overflow menu item in a page
+	overflowMenuItemId = Math.floor(Math.random() * 99999999999);
+
+	init() {
+		const options = this.model.getOptions();
+
+		// Grab the tooltip element
+		const holder = select(this.services.domUtils.getHolder());
+		const chartPrefix = Tools.getProperty(options, "style", "prefix");
+
+		this.overflowMenu = DOMUtils.appendOrSelect(
+			holder,
+			`div.${settings.prefix}--${chartPrefix}--overflowMenu`
+		);
+
+		this.overflowMenu.style("max-width", null);
+
+		// listen to show overflow menu event to render the overflow menu
+		this.services.events.addEventListener(
+			Events.Toolbar.SHOW_OVERFLOW_MENU,
+			() => {
+				this.overflowMenu.html(this.getOverflowMenuHTML());
+			}
+		);
+
+		// listen to hide overflow menu event to hide the overflow menu
+		this.services.events.addEventListener(
+			Events.Toolbar.HIDE_OVERFLOW_MENU,
+			() => {
+				this.overflowMenu.html(null);
+			}
+		);
+		// hide overflow menu if user clicks on somewhere in web page
+		document.body.addEventListener("click", () =>
+			this.showOverflowMenu(false)
+		);
+	}
+
+	render(animate = true) {
+		const isDataLoading = Tools.getProperty(
+			this.model.getOptions(),
+			"data",
+			"loading"
+		);
+
+		// size of toolbar button with background
+		const buttonSize = Configuration.toolbar.buttonSize;
+		const parentY = parseFloat(this.parent.node().getAttribute("y"));
+
+		const svg = this.getContainerSVG();
+
+		// TODO -- adjust toolbar Y position to align title component
+		// before layout component supports vertical alignment center
+		const VERTICAL_ALIGNMENT_Y = -6;
+		svg.attr(
+			"transform",
+			`translate(0, ${parentY + VERTICAL_ALIGNMENT_Y})`
+		);
+
+		const { width } = DOMUtils.getSVGElementSize(
+			this.services.domUtils.getMainSVG(),
+			{
+				useAttrs: true
+			}
+		);
+
+		// overflow menu width is 160px
+		// it's set by Carbon component
+		const overflowMenuWidth = 160;
+		// no good solution to get correct Toolbar position
+		// parent x doesn't work well
+		// assume the overflow icon has right alignment in layout
+		this.overflowMenuX = width - overflowMenuWidth;
+		this.overflowMenuY = parentY + VERTICAL_ALIGNMENT_Y + buttonSize;
+		const container = DOMUtils.appendOrSelect(svg, "svg.toolbar-container")
+			.attr("width", "100%")
+			.attr("height", Configuration.toolbar.height)
+			.attr("opacity", 1);
+
+		// clean children first
+		container.html(null);
```

Comment:
hmmmm... so the toolbar cleans everything everytime? Why not render more efficiently through d3?

Reply:
Yes, we need to clean buttons if data is loading, or update toolbar controls if configuration is changed.
We could use d3, but I don't think it could help to reduce the code complexity because the buttons and overflow menu items are more related to configurations instead of simple data, and for each configuration we need to create complex html component instead of SVG.
Pleasee let me know if you still think we should change to use d3.